### ISSUE TITLE
Fine tune tricks

### ISSUE THREAD

**Issue by timkang7** [Time-MoE/Time-MoE#61]
Hi, great work! I am trying to test TimeMoE-50M for fun. I try to fine-tune the model with the weather dataset from Time-300B/nature/weather.  Then I run this: "python main.py -d Time-300B/nature/weather --micro_batch_size=8". But the loss started with {'loss': **0.2304**, 'grad_norm': 0.11955317109823227, 'learning_rate': 5e-05, 'epoch': 0.0}   and end with {'loss': **0.2122**, 'grad_norm': 0.1316196322441101, 'learning_rate': 5e-05, 'epoch': 1.0} . So the performance on this weather looks still struggling. Could you give me some suggestions to fine tune it?

Looking forward to your guild line.

**Comment by timkang7** [comment]
Thanks for your reply！ I am a bit confused about the Time-300B data structure and couldn't find some relevant statements. Could you provide more details on this (e.g., Time-300B/nature/weather)?

In my approach, I use the first （Len(ds) - prediction_length） data points from a randomly selected sequence to predict the last prediction_length data points. I then compare the predictions with the actual values and compute the loss. The MAE and MSE loss values are mostly greater than 1, primarily ranging between 1 and 10, with some significantly exceeding 10. Does this seem reasonable?

Below is my test code:

prediction_length = 6
ds = TimeMoEDataset('Time-300B/nature/weather') 
seq_idx = random.randint(0, len(ds) - 1)
seq = ds[seq_idx]
seqs_t = torch.tensor(seq, dtype=torch.float32)

context_length = len(seq) - prediction_length
seqs = torch.from_numpy(seq[:context_length]).float().unsqueeze(0)
seqs = torch.tensor(seqs, dtype=torch.float32)

model = AutoModelForCausalLM.from_pretrained(
    'TimeMoE-50M',
    device_map="cuda",  # use "cpu" for CPU inference, and "cuda" for GPU inference.
    trust_remote_code=True,
)

# normalize seqs
mean, std = seqs.mean(dim=-1, keepdim=True), seqs.std(dim=-1, keepdim=True)
normed_seqs = (seqs - mean) / std

# Move tensors to CUDA before inference
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
normed_seqs = normed_seqs.to(device)
mean = mean.to(device)
std = std.to(device)

# forecast
output = model.generate(normed_seqs, max_new_tokens=prediction_length)  # shape is [batch_size, 12 + 6]
normed_predictions = output[:, -prediction_length:]  # shape is [batch_size, 6]

# inverse normalize
predictions = normed_predictions * std + mean

# Extract the last 6 values from seq (ground truth) for comparison
target = torch.from_numpy(seq[-prediction_length:]).float().to(device)
target = target.unsqueeze(0)

# Calculate MSE loss
mse_loss = torch.nn.MSELoss()
loss = mse_loss(predictions, target)
print(f"MSE Loss: {loss.item()}")

# Calculate MAE loss for reference
mae_loss = torch.mean(torch.abs(predictions - target))
print(f"MAE Loss: {mae_loss.item()}")

**Comment by Maple728** [comment]
The lengths of the sequences in the weather dataset can be very long and may exceed Time-MoE's maximum sequence length of 4096. When the sum of context_length and prediction_length exceeds 4096, the performance of Time-MoE may degrade. Therefore, please ensure that context_length + prediction_length <= 4096 when making predictions.

I hope my answer is helpful to you.

**Comment by timkang7** [comment]
Cool, much better after I limited the max length. Thanks for your job.
